### PR TITLE
feat: Added a thai context to the bidi and substitution features

### PR DESCRIPTION
**Pull request by rafallyczkowskiadylic**
## Description

feat: Added a thai context to the bidi and substitution features. 
Followed the current solution's convention to introduce thai script's features within the bidi implementation.

- added the thai word context using a unicode characters range for the thai range
- added the support for thai `liga`, `rlig`, `ccmp` substitution features
- updated the font default render options for thai scripts features support
- tests coverage

## Motivation and Context

Thai fonts / context had no support for glyph substitutions features that are required to render a valid thai contents:
- [Ligatures (liga)](https://learn.microsoft.com/en-us/typography/opentype/spec/features_ko#tag-liga)
- [Required Ligatures (rlig)](https://learn.microsoft.com/en-us/typography/opentype/spec/features_pt#tag-rlig)
- [Glyph Composition/Decomposition (ccmp) ](https://learn.microsoft.com/en-us/typography/opentype/spec/features_ae#tag-ccmp)

## How Has This Been Tested?

Added tests coverage to the `test/bidi.js` conducted agains the NotoSansThai-Medium font source ([SIL OFL](https://scripts.sil.org/cms/scripts/page.php?site_id=nrsi&id=OFL)) including: 
- should apply glyph composition feature
- should apply ligatures feature
- should apply required ligatures feature
- should match thai words in a given text (including mixed thai and english content)

## Screenshots (if appropriate):

image

## Thai font updated:

OFL font has been updated to add and test `rlig` feature as follows:
image (1)

with successful tests results WITH `rlig` support as:
image (2)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I did `npm run test` and all tests passed green (including code styling checks).
- [x] I have added tests to cover my changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the **README** accordingly.
- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

**Comment by ILOVEPIE** [comment]
@rafallyczkowskiadylic please rebase onto and create a pull request for the new active development fork [here](https://github.com/SABRE-JS/opentype.js)

File: src/bidi.js
Comment (ILOVEPIE) [inline src/bidi.js:160]: Is there a reason for this change?

File: src/bidi.js
Comment (ILOVEPIE) [inline src/bidi.js:172]: Is there a reason for this change?

**Review by ILOVEPIE — COMMENTED** [review]
Apart from those two changes at the beginning, this looks good. @rafallyczkowskiadylic can you explain these two changes please.

File: src/bidi.js
Comment (Connum) [reply src/bidi.js:160]: The hasFeatureEnabled() function was added with this PR and I guess the motivation was that it's easier to understand for contributors what's happening there. It's fine by me!

File: src/bidi.js
Comment (Connum) [reply src/bidi.js:172]: The hasFeatureEnabled() function was added with this PR and I guess the motivation was that it's easier to understand for contributors what's happening there. It's fine by me!

Comment (rafallyczkowskiadylic) [reply src/bidi.js:172]: This bit only got refactored, motivated:
- (primary) It was reasonable to not introduce another redundant code, and keep code DRY, with this trivial change
- (secondary) more comprehend code source

File: src/bidi.js
Comment (rafallyczkowskiadylic) [reply src/bidi.js:160]: This bit only got refactored, motivated:
- (primary) It was reasonable to not introduce another redundant code, and keep code DRY, with this trivial change
- (secondary) more comprehend code source

**Review by ILOVEPIE — COMMENTED** [review]
Do we have the licence to host NotoSansThai on the repository? We need to make sure of that before we approve this PR.

**Comment by Connum** [comment]
According to Google Fonts, it's licensed under the OFL: https://fonts.google.com/noto/specimen/Noto+Sans+Thai/about

File: src/features/thai/thaiLigatures.js
Comment (ILOVEPIE) [inline src/features/thai/thaiLigatures.js:39]: Need to add support for required ligatures.

File: src/font.js
Comment (ILOVEPIE) [inline src/font.js:283]: all font scripts are required to have rlig enabled by default. Please add rlig (Required Ligatures) feature.

File: test/bidi.js
Comment (ILOVEPIE) [inline test/bidi.js:103]: please add rlig (Required Ligatures).

**Review by ILOVEPIE — DISMISSED** [review]
The script is missing support for Required Ligatures (rlig). Please add this.

**Comment by ILOVEPIE** [comment]
liga is optional ligatures, rlig is required ligatures. rlig should never be disabled. The only reason we have the option to disable it is for tests and to make it easier for people to debug issues in code using opentype.js that are related to ligatures.

**Comment by ILOVEPIE** [comment]
@rafallyczkowskiadylic any updates on adding support for required ligatures?

**Comment by rafallyczkowskiadylic** [comment]
> @rafallyczkowskiadylic any updates on adding support for required ligatures?

@ILOVEPIE what is planned release date? 
I am planning to add them probably around this or next weekend but I can try to squeeze some time in sooner.

**Comment by rafallyczkowskiadylic** [comment]
Ok, no worries. I will engage for these change (both MRs) so we can all be happy with new features:)

**Comment by Connum** [comment]
@rafallyczkowskiadylic just a friendly nudge... :) Before the conflicts with main get too messy...

File: src/font.js
Comment (rafallyczkowskiadylic) [reply src/font.js]: Feature has been added

File: test/bidi.js
Comment (rafallyczkowskiadylic) [reply test/bidi.js]: Feature has been added

**Comment by rafallyczkowskiadylic** [comment]
@Connum @ILOVEPIE added `rlig` feature

File: src/features/thai/thaiLigatures.js
Comment (Connum) [inline src/features/thai/thaiLigatures.js:5]: I just noticed that the file extensions are missing, but are now required since we changed the build process and the library may be used in module form on the web, where we need the file extensions. I'm fixing it right now.

Comment (Connum) [reply src/features/thai/thaiLigatures.js:5]: @rafallyczkowskiadylic Would you mind allowing edits to the PR by maintainers, please?

Comment (rafallyczkowskiadylic) [reply src/features/thai/thaiLigatures.js]: Sure, added. Believe covered all

**Review by Connum — APPROVED** [review]
LGTM now! 🚀

File: src/features/thai/thaiLigatures.js
Comment (ILOVEPIE) [inline src/features/thai/thaiLigatures.js:26]: I don't generally recommend foreach as function calls can be very expensive when done in a loop, often a better solution is for-in loops or for loops. But we can fix this later.

File: src/features/thai/thaiRequiredLigatures.js
Comment (ILOVEPIE) [inline src/features/thai/thaiRequiredLigatures.js:26]: Same here with the forEach

File: src/features/thai/thaiGlyphComposition.js
Comment (ILOVEPIE) [inline src/features/thai/thaiGlyphComposition.js:26]: Foreach can be inefficent with long arrays as it requires a function call for each element in the array, and function calls can be slow if the compiler doesn't optimize them well. For-in or for loops are guaranteed to run fairly quickly however. This can be fixed later.

File: src/features/thai/thaiGlyphComposition.js
Comment (ILOVEPIE) [inline src/features/thai/thaiGlyphComposition.js:32]: See prior about forEach

**Review by ILOVEPIE — APPROVED** [review]
There's a few small performance things that i'm a little picky about, but other than that it looks good.

**Comment by Connum** [comment]
Regarding forEach performance, see also the ongoing discussion here: https://github.com/opentypejs/opentype.js/pull/581#discussion_r1125145076

**Comment by rafallyczkowskiadylic** [comment]
Resolved conflicts.

**Review by ILOVEPIE — APPROVED** [review]
Looks Good To Me.

**Pull request merged**
